src/parser/shared/modules/BuffCountGraph.tsx

```diff
@@ -0,0 +1,440 @@
+    this.ruleLineTrackers = this.castRuleSpecs().map((spec) =>
+      this._convertToInternalTracker(spec),
+    );
+    this.ruleLineTrackerLookup = this._generateLookup(this.ruleLineTrackers);
+    // create empty data list
+    this.graphData = [];
+    this.lastTimestamp = this.owner.fight.start_time; // timestamp of last item added to data
+
+    // setup listeners - spell list needs to filtered through a Set to remove duplicates
+    const buffSpellList = [...new Set(this.buffTrackers.flatMap((spec) => spec.spells))];
+    this.addEventListener(
+      Events.applybuff.by(SELECTED_PLAYER).spell(buffSpellList),
+      this.onBuffApplied,
+    );
+    this.addEventListener(
+      Events.removebuff.by(SELECTED_PLAYER).spell(buffSpellList),
+      this.onBuffRemoved,
+    );
+    this.addEventListener(
+      Events.applydebuff.by(SELECTED_PLAYER).spell(buffSpellList),
+      this.onBuffApplied,
+    );
+    this.addEventListener(
+      Events.removedebuff.by(SELECTED_PLAYER).spell(buffSpellList),
+      this.onBuffRemoved,
+    );
+
+    const castSpellList = [...new Set(this.ruleLineTrackers.flatMap((spec) => spec.spells))];
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(castSpellList), this.onCast);
+
+    this.addEventListener(Events.fightend, this.onFightEnd);
+  }
+
+  /**
+   * Implement to return the buffs to be graphed.
+   * This is a function instead of a field in the constructor so implementer can filter the specs
+   * based on player talents / legendaries / covenant.
+   */
+  abstract buffSpecs(): GraphedSpellSpec[];
+
+  /**
+   * Override to return casts to be graphed.
+   * This is a function instead of a field in the constructor so implementer can filter the specs
+   * based on player talents / legendaries / covenant.
+   */
+  castRuleSpecs(): GraphedSpellSpec[] {
+    return [];
+  }
+
+  /**
+   * Adds a vertical rule line for the given tracker at the given timestamp
+   * @param trackerName the name of the tracker this rule line is for -
+   *     will determine the line's color and its name in the legend
+   * @param timestamp the timestamp to add the rule
+   */
+  addRuleLine(trackerName: string, timestamp: number) {
+    const tracker = this.ruleLineTrackers.find((t) => t.name === trackerName);
+    if (!tracker) {
+      console.warn("Couldn't find tracker with name " + trackerName);
+      return;
+    }
+    this.graphData.push({
+      name: tracker.name,
+      type: 'rule',
+      timestamp,
+    });
+  }
+
+  onFightEnd() {
+    this._registerChange();
+  }
+
+  onBuffApplied(event: ApplyBuffEvent | ApplyDebuffEvent) {
+    this._onBuffChanged(event, 1);
+  }
+
+  onBuffRemoved(event: RemoveBuffEvent | RemoveDebuffEvent) {
+    this._onBuffChanged(event, -1);
+  }
+
+  _onBuffChanged(event: AbilityEvent<any>, change: number) {
+    const applicableTrackers = this.buffTrackerLookup[event.ability.guid];
+    if (applicableTrackers === undefined) {
```

Comment:
Could simplify this to `if(!applicableTrackers)` if I'm not mistaken

Reply:
tru tru